client.go

```diff
@@ -0,0 +1,289 @@
+package cfdns
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/simplesurance/cfdns/logs"
+	"github.com/simplesurance/cfdns/retry"
+)
+
+const (
+	retryFirsDelay   = 2 * time.Second
+	retryMaxDelay    = 30 * time.Second
+	retryFactor      = 2
+	retryMaxAttempts = 6
+	itemsPerPage     = 500
+)
+
+func NewClient(creds Credentials, options ...Option) *Client {
+	ret := Client{
+		cfg:   applyOptions(options...),
+		creds: creds,
+	}
+
+	return &ret
+}
+
+type Client struct {
+	cfg   *settings
+	creds Credentials
+}
+
+func runWithRetry[TREQ any, TRESP commonResponseSetter](
+	ctx context.Context,
+	logger *logs.Logger,
+	req *request[TREQ],
+) (
+	resp *response[TRESP],
+	_ error,
+) {
+	reterr := retry.ExpBackoff(ctx, logger, retryFirsDelay, retryMaxDelay,
+		retryFactor, retryMaxAttempts, func() error {
+			var err error
+			resp, err = runOnce[TREQ, TRESP](ctx, logger, req)
+			return err
+		})
+
+	return resp, reterr
+}
+
+// runOnce sends an HTTP request, parses and returns the response.
+// Permanent errors are wrapped with retry.PermanentError. Any error returned
+// from the server is wrapped with HTTPError. If the error is a valid
+// CloudFlare error, it is also wrapped with CloudFlareError.
+func runOnce[TREQ any, TRESP commonResponseSetter](
+	ctx context.Context,
+	logger *logs.Logger,
+	treq *request[TREQ],
+) (
+	tresp *response[TRESP],
+	err error,
+) {
+	if err = treq.client.cfg.ratelim.Wait(ctx); err != nil {
+		return
+	}
+
+	// request body
+	var reqBody []byte
+	if treq.body != nil {
+		reqBody, err = json.Marshal(treq.body)
+		if err != nil {
+			err = retry.PermanentError{Cause: err}
+			return
+		}
+	}
+
+	// request
+	req, err := http.NewRequestWithContext(ctx, treq.method, requestURL(treq),
+		bytes.NewReader(reqBody))
+	if err != nil {
+		err = retry.PermanentError{Cause: err}
+		return
+	}
+
+	// headers
+	mergeHeaders(req.Header, treq.headers)
+
+	// credentials
+	err = treq.client.creds.configure(
+		logger.SubLogger(logs.WithPrefix("Authorization")),
+		req)
+	if err != nil {
+		return // allow retry
```

Comment:
This would be useful if we add support for key rotation. If key rotation fails, it makes sense to try to rate it again before failing the http request.

Since it is not needed now I simplified the credentials interface, so now it does not take a logger or return an error.